tests/MMS/laplace3d/test-matrix.cxx

```diff
@@ -0,0 +1,75 @@
+#include <bout.hxx>
+
+#include <invert_laplace.hxx>
+#include <field_factory.hxx>
+#include <bout/constants.hxx>
+#include <derivs.hxx>
+#include "../../../src/invert/laplace/impls/petscamg/petscamg.hxx"
+
+Field3D this_Grad_perp2(const Field3D &f);
+Field3D this_Grad_perp_dot_Grad_perp(const Field3D &f, const Field3D &g);
+
+int main(int argc, char **argv) {
+  int init_err = BoutInitialise(argc, argv);
+  if (init_err < 0) {
+    return 0;
+  } else if (init_err > 0) {
+    return init_err;
+  }
+
+  Field3D D = 1.;
+  Field3D C1 = 1.;
+  Field3D C2 = 1.;
+  Field3D A = 0.;
+
+  // Create a Laplacian inversion solver
+  LaplacePetscAmg *lap = (LaplacePetscAmg*) Laplacian::create();
```

Comment:
I'm not overly keen on this. I guess because we need to access `multiplyAx` below?
Better options could be, ordered by my preference:
- Don't do `multiplyAx`. It sounds like this literally just does `A*x`, which should just be 0 anyway?
- Pull `multiplyAx` out of `LaplacePetscAmg` (probably not fun)
- Just create a `LaplacePetscAmg` on the stack (still requires relative import of private header)
- `auto lap = dynamic_cast<LaplacePetscAmg*>(...)`

Reply:
Once again, I didn't write this code (`LaplacePetscAmg` was written before I began working on this project, although I gather it still does not entirely work) so I'm perhaps not the best person to be fixing it. I'm willing to give it a go, however. What is the order of preference in your suggestions? Ascending or descending? (I assume they are decreasing order.)

Reply:
I did write the test, so maybe I should take responsibility for fixing it... unless you already fixed it @cmacmackin?

The purpose of this test was to check the matrix generation in `LaplacePetscAmg` by comparing the forward operation (`A*x = rhs`) to the analytical rhs (which was generated by sympy in `mms.py` and saved in the input file) and the forward operation implemented with BOUT++ operators.

Checking the forward operation makes it easier to find problems than in the solve. E.g. if there is an error in the boundary condition coefficient, it will appear just at that boundary instead of making the whole solution wrong.

Exposing `multiplyAx` is only needed for debugging, I can't think why it would be needed in production code. In principle if the full test from `tests/MMS/laplace3d/laplace.cxx` passes, then we don't need this one. I'd be inclined to keep it for the moment though with minimal changes, on the understanding that someone should clean up `LaplacePetscAmg` in future, but it is beyond the scope of this PR. So I'd be in favour of option 3 (create `LaplacePetscAmg` on the stack).

Reply:
I've updated the tests in `tests/MMS/laplace3d` so that they compile, but running `test-matrix` fails in `BoutFinalise()`. PETSc claims there is a problem in `Petsc_DelComm_Outer()` which is apparently being called from here for `test-matrix.cxx`
https://github.com/boutproject/BOUT-dev/blob/f24e39a597a1fbb94f1324ee1cd4a094514af9ee/src/mesh/impls/bout/boutmesh.cxx#L81-L82

The actual test `tests/MMS/laplace3d/runtest` (which runs `laplace`) passes!